File: backend/src/main/java/com/fitnexus/backend/service/PermissionService.java
package com.fitnexus.backend.service;

import com.fitnexus.backend.dto.CurrentUserPermissionsResponse;
import com.fitnexus.backend.dto.PermissionsMatrixResponse;
import com.fitnexus.backend.dto.PermissionMatrixUpdateRequest;
import com.fitnexus.backend.dto.PermissionPageDto;
import com.fitnexus.backend.dto.PagePermissionDto;
import com.fitnexus.backend.dto.RolePermissionsDto;
import com.fitnexus.backend.entity.PermissionPage;
import com.fitnexus.backend.entity.Role;
import com.fitnexus.backend.entity.RolePagePermission;
import com.fitnexus.backend.entity.Users;
import com.fitnexus.backend.exception.InvalidOperationException;
import com.fitnexus.backend.repository.PermissionPageRepository;
import com.fitnexus.backend.repository.RolePagePermissionRepository;
import com.fitnexus.backend.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Transactional
public class PermissionService {

    private final PermissionPageRepository permissionPageRepository;
    private final RolePagePermissionRepository rolePagePermissionRepository;
    private final UserRepository userRepository;

    public List<PermissionPageDto> getPages() {
        return permissionPageRepository.findAll().stream()
                .sorted(Comparator.comparing(PermissionPage::getSortOrder).thenComparing(PermissionPage::getLabel))
                .map(this::toPageDto)
                .toList();
    }

    public CurrentUserPermissionsResponse getCurrentUserPermissions(String email) {
        Users user = userRepository.findByEmail(email)
                .orElseThrow(() -> new InvalidOperationException("Current user not found"));
        Role role = user.getRole();
        return new CurrentUserPermissionsResponse(role, getPermissionsForRole(role));
    }

    public PermissionsMatrixResponse getMatrix() {
        List<PermissionPage> pageEntities = getSortedPages();
        List<PermissionPageDto> pages = pageEntities.stream().map(this::toPageDto).toList();
        List<RolePermissionsDto> roles = new ArrayList<>();

        for (Role role : Role.values()) {
            roles.add(new RolePermissionsDto(role, getPermissionsForRole(role, pageEntities)));
        }

        return new PermissionsMatrixResponse(pages, roles);
    }

    public List<PagePermissionDto> getPermissionsForRole(Role role) {
        return getPermissionsForRole(role, getSortedPages());
    }

    private List<PagePermissionDto> getPermissionsForRole(Role role, List<PermissionPage> pages) {
        Map<String, RolePagePermission> byPage = new HashMap<>();
        rolePagePermissionRepository.findByRole(role).forEach(item -> byPage.put(item.getPageKey(), item));

        return pages.stream()
                .map(page -> {
                    RolePagePermission rp = byPage.get(page.getPageKey());
                    if (rp != null) {
                        return new PagePermissionDto(page.getPageKey(), rp.isCanView(), rp.isCanCreate(), rp.isCanEdit(), rp.isCanDelete());
                    }
                    return new PagePermissionDto(page.getPageKey(), false, false, false, false);
                })
                .toList();
    }

    public void saveMatrix(PermissionMatrixUpdateRequest request) {
        if (request == null || request.getRoles() == null) {
            throw new IllegalArgumentException("Permission matrix is required");
        }

        Map<String, PermissionPage> pageLookup = new HashMap<>();
        permissionPageRepository.findAll().forEach(page -> pageLookup.put(page.getPageKey(), page));

        for (RolePermissionsDto roleDto : request.getRoles()) {
            if (roleDto == null || roleDto.getRole() == null) {
                continue;
            }

            List<PagePermissionDto> permissions = roleDto.getPermissions() == null ? List.of() : roleDto.getPermissions();
            Map<String, PagePermissionDto> submittedByPage = new HashMap<>();
            for (PagePermissionDto permission : permissions) {
                if (permission == null || permission.getPageKey() == null || !pageLookup.containsKey(permission.getPageKey())) {
                    continue;
                }
                submittedByPage.put(permission.getPageKey(), permission);
            }

            if (submittedByPage.isEmpty()) {
                rolePagePermissionRepository.deleteByRole(roleDto.getRole());
                continue;
            }

            rolePagePermissionRepository.deleteByRoleAndPageKeyNotIn(roleDto.getRole(), submittedByPage.keySet());

            Map<String, RolePagePermission> existingByPage = new HashMap<>();
            rolePagePermissionRepository.findByRole(roleDto.getRole())
                    .forEach(permission -> existingByPage.put(permission.getPageKey(), permission));

            List<RolePagePermission> entities = new ArrayList<>();
            for (PagePermissionDto permission : submittedByPage.values()) {
                RolePagePermission entity = existingByPage.getOrDefault(permission.getPageKey(), new RolePagePermission());
                entity.setRole(roleDto.getRole());
                entity.setPageKey(permission.getPageKey());
                entity.setCanView(permission.isCanView());
                entity.setCanCreate(permission.isCanCreate());
                entity.setCanEdit(permission.isCanEdit());
                entity.setCanDelete(permission.isCanDelete());
                entities.add(entity);
            }
            rolePagePermissionRepository.saveAll(entities);
        }
    }

    public boolean hasPermission(Role role, String pageKey, String action) {
        RolePagePermission permission = rolePagePermissionRepository.findByRoleAndPageKey(role, pageKey).orElse(null);
        if (permission == null) {
            return false;
        }
        return switch (action.toLowerCase()) {
            case "create" -> permission.isCanCreate();
            case "edit" -> permission.isCanEdit();
            case "delete" -> permission.isCanDelete();
            default -> permission.isCanView();
        };
    }

    private PermissionPageDto toPageDto(PermissionPage page) {
        return new PermissionPageDto(
                page.getPageKey(),
                page.getLabel(),
                page.getRoutePath(),
                page.getCategory(),
                page.getSortOrder()
        );
    }

    private List<PermissionPage> getSortedPages() {
        return permissionPageRepository.findAll().stream()
                .sorted(Comparator.comparing(PermissionPage::getSortOrder).thenComparing(PermissionPage::getLabel))
                .toList();
    }
}

File: backend-fastapi/app/services/permission_service.py
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.exceptions import InvalidOperationException
from app.models.permission import PermissionPage, RolePagePermission
from app.models.role import Role
from app.models.users import Users
from app.schemas.permissions import (
    CurrentUserPermissionsResponse,
    PagePermissionDto,
    PermissionMatrixUpdateRequest,
    PermissionPageDto,
    PermissionsMatrixResponse,
    RolePermissionsDto,
)


def _sorted_pages(db: Session) -> list[PermissionPage]:
    pages = list(db.scalars(select(PermissionPage)))
    return sorted(pages, key=lambda p: (p.sort_order, p.label))


def _to_page_dto(page: PermissionPage) -> PermissionPageDto:
    return PermissionPageDto(
        pageKey=page.page_key,
        label=page.label,
        routePath=page.route_path,
        category=page.category,
        sortOrder=page.sort_order,
    )


def get_permissions_for_role(db: Session, role: Role, pages: list[PermissionPage] | None = None) -> list[PagePermissionDto]:
    pages = pages if pages is not None else _sorted_pages(db)
    rows = db.scalars(select(RolePagePermission).where(RolePagePermission.role == role))
    by_page = {row.page_key: row for row in rows}

    result = []
    for page in pages:
        rp = by_page.get(page.page_key)
        if rp is not None:
            result.append(
                PagePermissionDto(
                    pageKey=page.page_key,
                    canView=rp.can_view,
                    canCreate=rp.can_create,
                    canEdit=rp.can_edit,
                    canDelete=rp.can_delete,
                )
            )
        else:
            result.append(
                PagePermissionDto(pageKey=page.page_key, canView=False, canCreate=False, canEdit=False, canDelete=False)
            )
    return result


def get_current_user_permissions(db: Session, email: str) -> CurrentUserPermissionsResponse:
    user = db.scalar(select(Users).where(Users.email == email))
    if user is None:
        raise InvalidOperationException("Current user not found")
    role = Role(user.role)
    return CurrentUserPermissionsResponse(role=role, permissions=get_permissions_for_role(db, role))


def get_matrix(db: Session) -> PermissionsMatrixResponse:
    page_entities = _sorted_pages(db)
    pages = [_to_page_dto(p) for p in page_entities]
    roles = [
        RolePermissionsDto(role=role, permissions=get_permissions_for_role(db, role, page_entities))
        for role in Role
    ]
    return PermissionsMatrixResponse(pages=pages, roles=roles)


def save_matrix(db: Session, request: PermissionMatrixUpdateRequest) -> None:
    if request is None or request.roles is None:
        raise ValueError("Permission matrix is required")

    page_keys = {p.page_key for p in db.scalars(select(PermissionPage))}

    for role_dto in request.roles:
        if role_dto is None or role_dto.role is None:
            continue

        submitted = {
            perm.pageKey: perm
            for perm in (role_dto.permissions or [])
            if perm is not None and perm.pageKey in page_keys
        }

        if not submitted:
            db.query(RolePagePermission).filter(RolePagePermission.role == role_dto.role).delete()
            continue

        db.query(RolePagePermission).filter(
            RolePagePermission.role == role_dto.role,
            RolePagePermission.page_key.notin_(submitted.keys()),
        ).delete(synchronize_session=False)

        existing = {
            row.page_key: row
            for row in db.scalars(select(RolePagePermission).where(RolePagePermission.role == role_dto.role))
        }

        for page_key, perm in submitted.items():
            entity = existing.get(page_key) or RolePagePermission(role=role_dto.role, page_key=page_key)
            entity.can_view = perm.canView
            entity.can_create = perm.canCreate
            entity.can_edit = perm.canEdit
            entity.can_delete = perm.canDelete
            db.add(entity)

    db.commit()


def has_permission(db: Session, role: Role, page_key: str, action: str) -> bool:
    permission = db.scalar(
        select(RolePagePermission).where(RolePagePermission.role == role, RolePagePermission.page_key == page_key)
    )
    if permission is None:
        return False
    action = (action or "").lower()
    if action == "create":
        return permission.can_create
    if action == "edit":
        return permission.can_edit
    if action == "delete":
        return permission.can_delete
    return permission.can_view
